api/interceptor/observer.go

```diff
@@ -0,0 +1,269 @@
+// request.
+func (r *RequestMetadata) StartAt() time.Time {
+	return r.startAt
+}
+
+// FinishedAt retrieves the timestamp indicating when the server completed
+// processing the request.
+func (r *RequestMetadata) FinishedAt() time.Time {
+	return r.finishedAt
+}
+
+// Duration returns the duration of the request processing.
+func (r *RequestMetadata) Duration() time.Duration {
+	return r.duration
+}
+
+// Procedure returns the RPC name.
+func (r *RequestMetadata) Procedure() string {
+	return r.procedure
+}
+
+// RequestURI returns the HTTP request uri.
+func (r *RequestMetadata) RequestURI() string {
+	return r.requestURI
+}
+
+// Method returns the HTTP method.
+func (r *RequestMetadata) Method() string {
+	return r.method
+}
+
+// Protocol returns the protocol that was used (e.g. grpc, connect, http).
+func (r *RequestMetadata) Protocol() string {
+	return r.protocol
+}
+
+// PeerAddress returns the IP address of the requester.
+func (r *RequestMetadata) PeerAddress() string {
+	return r.peerAddress
+}
+
+// MessagesReceived returns the number of messages received
+// in a streaming request.
+func (r *RequestMetadata) MessagesReceived() int {
+	return r.messagesReceived
+}
+
+// MessagesSent returns the number of messages sent
+// in a streaming request.
+func (r *RequestMetadata) MessagesSent() int {
+	return r.messagesSent
+}
+
+// BytesReceived is the number of bytes we read via HTTP.
+func (r *RequestMetadata) BytesReceived() int64 {
+	return r.bytesReceived
+}
+
+// BytesSent is the number of bytes we sent via HTTP.
+func (r *RequestMetadata) BytesSent() int64 {
+	return r.bytesSent
+}
+
+// HTTPStatusCode returns the HTTP status code. This is expected to be 200 for
+// most gRPC and connectrpc requests.
+func (r *RequestMetadata) HTTPStatusCode() int {
+	return r.httpStatusCode
+}
+
+// StatusCode returns the response status code in a string presentation.
+// If there was no error the status code "ok" will be returned.
+func (r *RequestMetadata) StatusCode() string {
+	if r.procedure == "none" && r.protocol == "http" {
+		switch r.httpStatusCode {
+		case http.StatusOK:
+			return "ok"
+		case http.StatusNotFound:
+			return connect.CodeNotFound.String()
+		case http.StatusRequestTimeout:
+			return connect.CodeCanceled.String()
+		case http.StatusBadRequest:
+			return connect.CodeInvalidArgument.String()
+		default:
+			return connect.CodeInternal.String()
+		}
+	}
+
+	if r.err == nil {
+		return "ok"
+	}
+	return connect.CodeOf(r.err).String()
+}
+
+// Err returns the error that was returned by a handler.
+func (r *RequestMetadata) Err() error {
+	return r.err
+}
+
+func (r *RequestMetadata) end() {
+	r.finishedAt = time.Now()
+	r.duration = time.Since(r.startAt)
+}
+
+type requestMetadataCtxKey struct{}
+
+func setRequestMetadata(ctx context.Context, rm *RequestMetadata) context.Context {
+	return context.WithValue(ctx, requestMetadataCtxKey{}, rm)
+}
+
+func getRequestMetadata(ctx context.Context) *RequestMetadata {
```

Comment:
Shouldn't this be exported, so callers can get the request metadata? i.e. they add a middlewaree after this one.

Reply:
If we export it, the callers must be aware when to use it though, otherwise it'll panic. I'd keep it unexported for now until we have this usecase and then we can still export it. Do you plan to do something like this already?